### ISSUE TITLE
Installer

### ISSUE THREAD

**Issue by HoeblingerDaniel** [TcNobo/TcNo-Acc-Switcher#36]
Could you make an installer that also puts it into the start menu? 
I wanted to show my friends but it's quite complicated for nontech people.

**Comment by TcNobo** [comment]
Yes, this is planned... Although I've been delaying it for 2 reasons:
1. There is no auto-installer to grab the 3 runtimes you had to download... Having to make my own, but I've been sidetracked with bugfixes :/
2. Some versions in this beta have been mega borked and won't update at all. Having an installer might make it a little more confusing to replace files if necessary, although the bulk of these major issues have been passed.

tldr I should be adding it sometime soon.

**Comment by kotx** [comment]
It might be worth it to port to .NET 5 and use MSIX packages.

**Comment by TcNobo** [comment]
The Beta is .NET 5 already. Blazor/C# in .NET 5 ~
I looked a little into MSIX packages and sharing files also through the Microsft store. The store seems very limited, at least for this project.
I originally asked on Reddit about cheaper code signing ([See post here](https://www.reddit.com/r/csharp/comments/nuh1ua/who_can_i_trust_code_signing_for_an_opensource/)). TLDR I went with buying a standard cert for this project. Will use it elsewhere too. Should have it done sometime soon.

The longest response thread talks about permissions and the rest on the Microsoft store. Read through it and it seems OK, just that no files would be user-editable in the main program folder, so I'd have to move everything (LoginCache, themes...) into a Documents folder for example. Not out of the scope of this project, just some extra effort.
I assume you're talking about just MSIX packages not necessarily destined for the MS store? I assume those would have user-editable files in the install folder and not as sandboxed as apps on the store.

For now, at least, I'll definitely build a normal NSIS installer as soon as I'm done with the more user-friendly installer that will grab all 3 runtimes and open the installers for users. People almost always miss one of the 3, breaking the entire project.
[Only really caught up on grabbing download links. I'm split with just using existing links, or grabbing them from HTML at runtime. I don't know if the randomly generated looking direct download links include an expiry, limit or anything along those lines]

**Comment by kotx** [comment]
Thanks for the quick response- didn't realize beta was already in .NET 5, that's great. MSIX packages do have user-editable files IIRC, just that the path it installs to is not user-customizable. It's hard to find since the folder name has a GUID in it, but you should be able to open explorer in the folder from the app. 
Moving everything to Documents (or even a user-customizable folder) might be a good idea if going down this route.

The WebView2 runtime should always be available here https://go.microsoft.com/fwlink/p/?LinkId=2124703 (the same link in your README), so that should be ok to programmatically download and run. I don't know about the .NET 5 runtime though.

You can always publish [self-contained applications](https://docs.microsoft.com/en-us/dotnet/core/deploying/#publish-self-contained) instead of requiring a global runtime.

**Comment by TcNobo** [comment]
I'd rather keep as small a download as possible if people need to redownload.
In better news, I found a [dotnet-install](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) script from Microsoft. Looking at the ps1 code, I got these links:
https://dotnetcli.azureedge.net/dotnet/WindowsDesktop/5.0.7/windowsdesktop-runtime-5.0.7-win-x64.zip
https://dotnetcli.azureedge.net/dotnet/aspnetcore/Runtime/5.0.7/aspnetcore-runtime-5.0.7-win-x64.zip

Replace .zip with .exe, and it downloads the latest installer. Going to double-check now on a VM if this works, but the Hosting Bundle includes .NET 5 (What I tell people to download). This ASPNetCore runtime doesn't, so it's only 7.9MB instead of 63MB.
If it works it should take it from:
-  WebView2 Runtime (Though I'm sure it downloads more when run) - 1,70 MB => 1,70 MB (Unchanged)
- .NET 5.0 Desktop Runtime - 52,3 MB => 52,3 MB (Unchanged)
- Hosting Bundle - 63,2 MB => 7,90 MB [As it drops a second copy of .NET 5]
So from 117,2 MB to 61,9. Half the download size, and it can be automated. Now just to figure out the `/S` or whatever for the installer to skip prompts.

**Comment by TeleAgentc1** [comment]
how to run this app developer ?

**Comment by TcNobo** [comment]
@TeleAgentc1 Download all 3 runtimes, as shown here: https://github.com/TcNobo/TcNo-Acc-Switcher#required-runtimes-download-and-install
Download and unzip the latest release from here: https://tcno.co/Projects/AccSwitcher/api/latest
Run `TcNo-Acc-Switcher.exe`

**Comment by TcNobo** [comment]
The runtime installer is complete, and I will document the "First run" procedure.
When a missing runtime(s) are found: 
![image](https://user-images.githubusercontent.com/10319195/122048908-2317d700-cde2-11eb-831e-a51ad271a7b1.png)


After everything is installed: 
![image](https://user-images.githubusercontent.com/10319195/122048189-42facb00-cde1-11eb-8d1d-9303d753683a.png)

**Comment by TcNobo** [comment]
Looks like we're steps closer to an MSIX package...
The cache for the webview needs to be located somewhere other than Program Files, otherwise, this error appears:
![image](https://user-images.githubusercontent.com/10319195/122069729-b575a600-cdf5-11eb-9c59-9807641e8339.png)

[Info on this error](https://github.com/MicrosoftEdge/WebView2Feedback/issues/297).
Time for some refactoring to start using the Documents directory now.
Sidenote - I've also submitted more documents to Sectigo, so signed binaries are a short while away - hopefully.

**Comment by TcNobo** [comment]
Everything seems good. Tested on a fresh Windows 10 VM. Installed all platforms, signed in, installed TcNo Account Switcher, everything went smooth. Testing led to a few interesting bugs that have since been fixed. Everything should be in working order and ready for release...

For now, if you'd like to test it and give me feedback here or via Discord, please do.
Download the installer [HERE - https://tcno.co/dl/TcNo%20Account%20Switcher%20-%20Installer.exe](https://tcno.co/dl/TcNo%20Account%20Switcher%20-%20Installer.exe)

**Comment by kotx** [comment]
I tried out the installer- seems to be all good, though the only runtime I was missing was the WebView2 runtime so I couldn't test the other runtimes. Works great though!

**Comment by TcNobo** [comment]
Alright, the installer is now public and available with releases through GitHub and the Discord server. Should it need any changes, do let me know.
It adds the files, uninstaller. registered protocol support into Windows and finally installs all required runtimes as well.

I'm in the process of editing the Wiki to better show the new install process, and the main page as well. When it's finished I'll also make an install guide video to feature.

I'll leave this thread open for a short while in case more info or additions are necessary. Do let me know if things need to be changed at all.
